src/aggregators/prioritytable.jl

```diff
@@ -312,3 +318,155 @@ function sample(pt::PriorityTable, priorities, rng = DEFAULT_RNG)
     # sample element within the group
     @inbounds sample(groups[gid], priorities, rng)
 end
+
+##########################
+### Routines for CCNRM ###
+##########################
+
+mutable struct TimeGrouper{T <: Number}
+    mintime::T
+    timestep::T
+end
+
+@inline function (t::TimeGrouper{T})(time::T) where {T}
+    return floor(Int, (time - t.mintime) / t.timestep) + 1
+end
+
+mutable struct PriorityTimeTable{T, F <: Int}
+    groups::Vector{PriorityGroup{T, Vector{F}}}
+    pidtogroup::Vector{Tuple{F, F}}
+    times::Vector{T}
+    timegrouper::TimeGrouper{T}
+    minbin::F
+    steps::F # TODO: For adaptive rebuilding. 
+    maxtime::T
+end
+
+# Construct the time table with the default optimal bin width and number of bins. 
+# DEFAULT NUMBINS: 20 * √length(times)
+# DEFAULT BINWIDTH: 16 / sum(propensities)
+function PriorityTimeTable(times::AbstractVector, mintime, timestep)
+    numbins = floor(Int64, 20 * sqrt(length(times)))
+    maxtime = mintime + numbins * timestep
+
+    pidtype = typeof(numbins)
+    ptype = eltype(times)
+    groups = Vector{PriorityGroup{ptype, Vector{pidtype}}}()
+    pidtogroup = Vector{Tuple{Int, Int}}(undef, length(times))
+
+    ttgdata = TimeGrouper{ptype}(mintime, timestep)
+    # Create the groups, [t_min, t_min + τ), [t_min + τ, t_min + 2τ)...
+    for i in 1:numbins
+        push!(groups, PriorityGroup{pidtype}(mintime + i * timestep))
+    end
+
+    ptt = PriorityTimeTable(
+        groups, pidtogroup, times, ttgdata, zero(pidtype), zero(pidtype), maxtime)
+    # Insert priority ids into the groups
+    for (pid, time) in enumerate(times)
+        if time > maxtime
+            pidtogroup[pid] = (0, 0)
+            continue
+        end
+        insert!(ptt, pid, time)
+    end
```

Comment:
Yeah, maybe it would be better to uniformly distribute them among bins.

But for now you can leave as is -- we do the same in DirectCR I think anyways, so it should probably also get reexamined.